File: practice_java_program/notpad/PossibleStringPermutationsUsingRecursion.java
public class PossibleStringPermutationsUsingRecursion{

    public static void main(String[] args) {
        String str = "ABC";
        permute(str, "");
    }

    public static void permute(String str, String result) {

        // Base case
        if (str.length() == 0) {
            System.out.println(result);
            return;
        }

        for (int i = 0; i < str.length(); i++) {

            char current = str.charAt(i);

            // Remaining string after removing current character
            String remaining = str.substring(0, i) + str.substring(i + 1);

            // Recursive call
            permute(remaining, result + current);
        }
    }
}

File: python_oriented_interview/notpad/PossibleStringPermutationsUsingRecursion.py
def permute(str_val, result):
    if len(str_val) == 0:
        print(result)
        return

    for i in range(len(str_val)):
        current = str_val[i]
        remaining = str_val[:i] + str_val[i+1:]
        permute(remaining, result + current)

if __name__ == "__main__":
    str_val = "ABC"
    permute(str_val, "")
